kimchi/src/circuits/polynomials/not.rs

```diff
@@ -0,0 +1,235 @@
+            Wire::for_row(zero_row),
+            GenericGateSpec::Const(F::zero()),
+            None,
+        ));
+        gates.extend(not_gates);
+        // check fin_in1, fin_in2, fin_out are zero
+        gates.connect_cell_pair((zero_row, 0), (zero_row, 1));
+        gates.connect_cell_pair((zero_row, 0), (zero_row, 2));
+        // Integration
+        gates.connect_cell_pair((pub_row, 0), (new_row, 1)); // input2 of xor is all ones
+
+        gates.len()
+    }
+}
+
+/// Extend a Not witness for less than 255 bits (native field)
+/// Input: full witness, first input and optional bit length
+pub fn extend_not_xor_witness<F: PrimeField>(
+    witness: &mut [Vec<F>; COLUMNS],
+    input: F,
+    bits: Option<usize>,
+) {
+    let input = input.to_biguint();
+    let output = BigUint::bitwise_not(&input, bits);
+    let bits = max(input.bitlen(), bits.unwrap_or(0));
+    let mut not_witness: [Vec<F>; COLUMNS] =
+        array::from_fn(|_| vec![F::zero(); num_xors(bits) + 1]);
+    init_xor(
+        &mut not_witness,
+        0,
+        bits,
+        (
+            F::from_biguint(&input).unwrap(),
+            F::from(2u8).pow(&[bits as u64]) - F::one(),
+            F::from_biguint(&output).unwrap(),
+        ),
+    );
+
+    for col in 0..COLUMNS {
+        witness[col].extend(not_witness[col].iter());
+    }
+}
+
+/// Create a Not witness for less than 255 bits (native field) starting at row 0
+/// Input: first input and optional bit length
+pub fn create_not_xor_witness<F: PrimeField>(input: F, bits: Option<usize>) -> [Vec<F>; COLUMNS] {
+    let mut witness: [Vec<F>; COLUMNS] = array::from_fn(|_| vec![F::zero(); 1]);
+    let input_big = input.to_biguint();
+    let real_bits = max(input_big.bitlen(), bits.unwrap_or(0));
+    witness[0][0] = F::from(2u8).pow(&[real_bits as u64]) - F::one();
+    extend_not_xor_witness(&mut witness, input, bits);
+    witness
+}
+
+/// Creates as many negations as the number of inputs. The inputs must fit in the native field.
+/// We start at the row 0 using generic gates to perform the negations.
+/// Input: a vector of words to be negated, and the number of bits (all the same)
+/// Panics if the bits length is too small for the inputs
+pub fn create_not_gnrc_witness<F: PrimeField>(inputs: &[F], bits: usize) -> [Vec<F>; COLUMNS] {
+    let mut witness: [Vec<F>; COLUMNS] = array::from_fn(|_| vec![F::zero(); 1]);
+    witness[0][0] = F::from(2u8).pow(&[bits as u64]) - F::one();
+    extend_not_gnrc_witness(&mut witness, inputs, bits);
+    witness
+}
+
+/// Extends negation witnesses from generic gate, assuming the input witness already contains
+/// public input rows holding the 2^bits-1 value.
+/// Input: a vector of words to be negated, and the number of bits (all the same)
+/// Panics if the bits length is too small for the inputs
+/// INTEGRATION: Set public input of bits in public generic gate
+/// TODO: `witness[0][pub] = 2^bits - 1`
```

Comment:
Well, it is not a TODO, but like a NOTE. Like, hey, remember that somewhere in the `pub` row, the column 0 needs to be set to the all one word.

Reply:
I think you need to write that explicitly.  So, instead of just saying
NOTE: `witness[0][pub] = 2^bits - 1`

Say, 

"NOTE: The caller must ensure that somewhere in the pub row, column zero must be set to `2^bits - 1`"

Reply:
Okay, well that is done in the INTEGRATION note, that is precisely what I meant. But okay, I will merge these two in one.